src/helpers/binding.ts

```diff
@@ -0,0 +1,163 @@
+const placeholderRegExp =
+  /["'`](?:(?<=")[^"\\]*(?:\\.[^"\\]*)*"|(?<=')[^'\\]*(?:\\.[^'\\]*)*'|[^`]*`)|(\?\d*|[:@$][a-zA-Z_]\w+)/gmu;
+
+function isPlainObject(obj: any): obj is Record<string, BaseType> {
+  if (typeof obj !== "object" || obj === null) return false;
+  const proto = Object.getPrototypeOf(obj);
+  return proto !== null && Object.getPrototypeOf(proto) === null;
+}
+
+function bindString(param: string, quoteEscaper = "''"): string {
+  return "'" + param.replace(/'/g, quoteEscaper) + "'";
+}
+
+function bindBoolean(param: boolean): string {
+  return Number(param).toString();
+}
+
+function bindNumber(param: number | bigint): string {
+  return param.toString();
+}
+
+function bindDate(param: Date): string {
+  return param.valueOf().toString();
+}
+
+function bindBytes(param: Uint8Array): string {
+  const hex = param.reduce((t, x) => t + x.toString(16).padStart(2, "0"), "");
+  return `X'${hex}'`;
+}
+
+function bindObject(param: any): string {
+  return bindString(JSON.stringify(param));
+}
+
+function bindNull(_param: undefined | null): string {
+  return "NULL";
+}
+
+function bindToString(param: any): string {
+  return bindString(String(param));
+}
+
+interface SQL {
+  toSQL: () => string;
+}
+
+function isSQL(param: any): param is SQL {
+  return typeof param.toSQL === "function";
+}
+
+function bindToSQL(param: SQL): string {
+  return param.toSQL();
+}
+
+function bindValue(param: BaseType): string {
+  switch (typeof param) {
+    case "bigint":
+    case "number":
+      return bindNumber(param);
+    case "boolean":
+      return bindBoolean(param);
+    case "string":
+      return bindString(param);
+    case "undefined":
+      return bindNull(param);
+    case "object":
+      if (param instanceof Date) {
+        return bindDate(param);
+      } else if (param instanceof Uint8Array) {
+        return bindBytes(param);
+      } else if (param == null) {
+        return bindNull(param);
+      } else if (isPlainObject(param)) {
+        return bindObject(param);
+      } else if (isSQL(param)) {
+        return bindToSQL(param);
+        /* c8 ignore next 3 */
+      } else {
+        return bindToString(param);
+      }
+    default:
+      return bindToString(param);
+  }
+}
+
+export type BaseType =
```

Comment:
I actually define the types we support here. This _could_ probably be left as any, but this typed version is nice because it'll catch weird things for us. If using from JavaScript, folks can get around this, so it isn't really saving us from runtime weirdness.